java/arcs/core/analysis/InformationFlow.kt

```diff
@@ -270,11 +272,47 @@ class InformationFlow private constructor(
     }
 
     /** Returns the [AccessPath.Selector] part of [AccessPath] instances for this [Schema]. */
-    private fun Schema.accessPathSelectors(): Set<List<AccessPath.Selector>> {
-        return (
-            fields.singletons.keys.map { listOf(AccessPath.Selector.Field(it)) } +
-            fields.collections.keys.map { listOf(AccessPath.Selector.Field(it)) }
-        ).toSet()
+    private fun Schema.accessPathSelectors(
+        prefix: List<AccessPath.Selector>
+    ): Set<List<AccessPath.Selector>> {
+        return fields.singletons.accessPathSelectors(prefix) +
+        fields.collections.accessPathSelectors(prefix)
+    }
+
+    private fun Map<FieldName, FieldType>.accessPathSelectors(
+        prefix: List<AccessPath.Selector>
+    ): Set<List<AccessPath.Selector>> {
+        return asSequence().fold(emptySet<List<AccessPath.Selector>>()) { acc, (name, type) ->
+            acc + type.accessPathSelectors(prefix + listOf(AccessPath.Selector.Field(name)))
+        }.toSet()
+    }
+
+    private fun FieldType.accessPathSelectors(
+        prefix: List<AccessPath.Selector>
+    ): Set<List<AccessPath.Selector>> {
+        return when (tag) {
+            // For primitives and references we don't go down to collect access paths.
+            FieldType.Tag.Primitive,
+            FieldType.Tag.EntityRef -> setOf(prefix)
```

Comment:
It is a bit tricky with references as we can have cycles in access paths.  For instance, consider a linked list data structure. The set of access paths is actually infinite: `l.next`, `l.next.next`, `l.next.next.next`, `...`

The common approach to deal with it is to limit the length of the access paths. If the length is limited to 3, we will only have access paths `l.next`, `l.next.next`, and `l.next.next.next` in the linked list example. `l.next.next.next` will capture the labels of all paths >= 3.

The implication of this approximation is that we can flag a recipe as violating policy even if it may not do so at runtime. That is, DFA is sound, but will erroneously flag a recipe as invalid. (It will never flag an invalid recipe as valid.)

The example above should work by summarizing the labels of `input.foo.field2` under `input.foo`, but I noticed a bug in the `edgeTransfer` function for longer paths involving references. It does not affect any of the exisiting use cases. I will send out a separate PR to deal with it.

Reply:
I don't believe we support cyclic references (yet)

Reply:
Won't the following schemas create cyclic references?

```
schema Particle
   name: Text
   handle: &Handle

schema Handle
  name: Text
  particle: &Particle
```

Reply:
I just tried this and the manifest parser throws the following error:

`Error: Could not resolve type reference to type name 'Handle'`

Is there any other way to create cyclic references?

Reply:
No, they're not supported at all.

Reply:
OK, I added support for going down entity reference fields as well. I will add a separate test in typescript that errors out when there are cycles.